# Basics of Python - Part 3 - Conditionals and Loops

In this module, we will look at the conditionals and for and while loops briefly. Why conditionals? Conditional statements are boolen statements helping us to check some given conditions and act accordingly. For example, we tell the computer do this:

If this statement is true, then perform b.

On the other hand, for and while statements helps us to do the repetitive statements in an efficient way. If we want to do the same repetitive task over and over by using values similar to each other, we could use a for (or while) statement instead of typinng the same thing over an over. By doing this, we write a more clean, neat code while saving a lot of time.

Here is the related documentation about these: https://docs.python.org/3/tutorial/controlflow.html 

## If statements

If statements are used to check conditional arguments. Checking a condition means using comparison operators to find out if the statement is True or False. So, we use ==, !=, <, <=, > and >= to check whether a statement is True or False (Please refer to Part 1 for the comparison operators). 

* If we are checking several conditions at the same time, we use if, elif and else as follows. (If we already used one if statement, the second condition must be written with elif (shortcut for else if)) 

* Here is how an if statement works: Computer checks the first condition; if it is True, it will give you that output. If not True, it goes to the second condition. If the second one is True, that will produce the output. If that is also not True, it will check the next statement and so on.

* We write the last statement with else. This is optional. Note that else statement doesn't have any conditional argument. This means if all other conditions are not satisfied, this is what you want your computer to do. 

* Indentation is very important when writing these statements. It is 4 spaces from left. Also, don't forget semicolon : at the end of each statement. Otherwise, it will give a SyntaxError.

Practice these below with the following codes. Note that I changed the value of apples each time so that it prints a different portion of the code.

    

In [1]:
apples=12

if apples==12:
    print('Since apples=12 the first portion of the code is satisfied and the code prints this.')
elif apples !=10:
    print("Even if this is True, the code stopped at the first step since that condition is satisfied. You won't see this.")
else:
    print("The first condition is already True, so you won't see this.")



Since apples=12 the first portion of the code is satisfied and the code prints this.


In [2]:
apples=11

if apples==12:
    print("This is skipped since apples is not equal to 12.")
elif apples !=10:
    print("You are seeing this because first condition is not satisfied and this one is satisfied.")
else:
    print("The second condition is already True, so you won't see this.")



You are seeing this because first condition is not satisfied and this one is satisfied.


In [3]:
apples=10

if apples==12:
    print("This is skipped since apples is not equal to 12.")
elif apples !=10:
    print("This is skipped since apples is equal to 10.")
else:
    print('Since the other two conditions are not satisfied, this is what it will print next.')


Since the other two conditions are not satisfied, this is what it will print next.


In addition, we could combine several conditions with logical operators 'and', 'or' and 'not'. 'and' means both conditions must be True for that code portion to run; 'or' statement means, if one of them True, that code portion would still run. 

In [4]:
apples=12
oranges=10
pears=14


if apples<oranges or apples<pears:
    print("Since both of these conditions are satisfied, you are seeing this output.")
else:
    print("If you are seeing this output, it means one of the conditions in the if statement is not satisfied.")

#If you want to see how it changes when we change 'and' to 'or', please uncomment the code below and run. 
#Seeing both outputs at the same time might be confusing, so you may want to comment the code above 
#by using triple quotes to see only one output.



if apples<oranges or apples<pears:
    print("If you are seeing this output, it means that one of the conditions must be satisfied.")
else:
    print("If you are seeing this output, it means none of the conditions is satisfied.")


print('------------')

Since both of these conditions are satisfied, you are seeing this output.
If you are seeing this output, it means that one of the conditions must be satisfied.
------------


Here is a cool tip: If you are unsure of what your code is doing, have a couple of print statement in each part so that you
will see when it is running, or how it is running etc. I will be using this trick below to explain for statements, as well.

## For statements

In the example below, the task is simple: Multiply numbers in the given list by 3 and write each result separately. The list given below has only 4 elements. To perform this task one by one, we have to write print(my_list[position]*3) 4 times to produce the results that we want. Imagine that we have 1000 elements! There must be a better way of doing this simple task. So, instead, we write a two line code by using a for loop (which is another control flow statement) and let it iterate through the list:


In [1]:
my_list=[1,2,3,4]

print(my_list[0]*3)
print(my_list[1]*3)
print(my_list[2]*3)
print(my_list[3]*3)

print('--------------------')

'''Just for fun, uncomment the code below and see what it prints.'''

#print(my_list*3)

print('--------------------')

for n in my_list:
    print(n*3)    

3
6
9
12
--------------------
--------------------
3
6
9
12


## Example 1:

Find all numbers between 1 and 40 that are divisible by 3?

We solve this in the following way:

* First, we create an empty list. 
* By using a for loop, we add numbers satisfiying the condition to the list.
* Since there is a condition to be checked, we have to use an if statement inside the for statement. Therefore, we have to nest it  in the following way. Note that every time, indentation=4 spaces.
* The condition given is that the number should be divisible by 3. So, we will use %3, which gives the remainder. If a number is divisible by 3, the remainder becomes 0. 


In [6]:

divisible_3=[] 

for n in range(1,40):
    if n%3==0:
        divisible_3.append(n)

#Now, let's see the list of all these numbers.
print(divisible_3)


[3, 6, 9, 12, 15, 18, 21, 24, 27, 30, 33, 36, 39]


## Exercise:

Find all numbers in list lst=[5, 10, 15, 20, 25, 30, 35, 40, 45,50,55,60,65,70,75,80] that are divisible by both 5 and 7.

Answer: [35,70] 

In [7]:
lst=[5, 10, 15, 20, 25, 30, 35, 40, 45,50,55,60,65,70,75,80]



## Example 2:

Let's consider all numbers between 10 and 20. You can use range(10,21) for this list. (Please don't forget: indexing starts with 0 so you have to have 11 to have 10 elements.)

- Multiply all even numbers between 10 and 20 by 10 and add them to a new list, called lst.
- Multiply all numbers that are divisible by 3 by 100, add those to lst, as well.

(a) How many elements are in lst?

(b) What is the sum of elements in lst?


In [8]:
lst1=[]

for n in range(10,21):
    if n%2==0:
        lst1.append(n*10)

for n in range(10,21):
    if n%3==0:
        lst1.append(n*100)

#Now, let's see the list of all these numbers.

print(lst1)

print('----------')

#We count them by len() and find the sum by sum():

print('(a) There are {} elements in lst.'.format(len(lst1)))

print('(b) The sum of all elements in lst is {}.'.format(sum(lst1)))


[100, 120, 140, 160, 180, 200, 1200, 1500, 1800]
----------
(a) There are 9 elements in lst.
(b) The sum of all elements in lst is 5400.


In the code above, we used two for loops since two conditions were given. So, the computer performs the first one, add items to the list, then do the second one. The list is constructed accordingly.

Instead of using two for loops, we could do this with one for loop. However, we have to use two if statements as two conditions are given separately. Please see the remark below to see what happens if we use elif instead of if. 

In addition, this will produce a (differently ordered) list since the for loop runs through 10 to 21 one by one.

In [9]:
lst2=[]

for n in range(10,21):
    if n%2==0:
        lst2.append(n*10)
    if n%3==0:
        lst2.append(n*100)

#Now, let's see the list of all these numbers.

print(lst2)

print('----------')

#We count them by len() and find the sum by sum():

print('(a) There are {} elements in lst.'.format(len(lst2)))

print('(b) The sum of all elements in lst is {}.'.format(sum(lst2)))


[100, 120, 1200, 140, 1500, 160, 180, 1800, 200]
----------
(a) There are 9 elements in lst.
(b) The sum of all elements in lst is 5400.


## Remark 1: 

(1) Note that in the code above, we used two if statements in the for statement. It was because of the instructions given. In addition, note that for loop runs one by one. Therefore, since 12 is divisible by 2 and 3, we have 120 (because it was divisible by 2) and 1200 (because it was divisible by 3). Since there are two if statements, we see both results when it runs 3.

However, if you use elif instead of if, it will skip the second part (with elif) since the first portion of the code is true. Therefore, the answer would be incorrect. Please see the code below:

(2) In addition, note that the computer does not perform this in the way given in the statement. Since the code goes through the list one by one, we see 1500 (a result of the second statement given) added in the code when it runs 15.

In [17]:
lst_with_elif=[]

for n in range(10,21):
    if n%2==0:
        lst_with_elif.append(n*10)
    elif n%3==0:
        lst_with_elif.append(n*100)

print(lst_with_elif)


[100, 120, 140, 1500, 160, 180, 200]


## Remark 2:

Let's look at how computer understands lists that we found:

In [18]:
print('This is the first list: {}'.format(lst1))
print('This is the second list: {}'.format(lst2))

print('----------')

print(lst1==lst2)


This is the first list: [100, 120, 140, 160, 180, 200, 1200, 1500, 1800]
This is the second list: [100, 120, 1200, 140, 1500, 160, 180, 1800, 200]
----------
False


As you can see, these two lists are different to the computer even though they contain the same elements. Since lists are ordered, computer understands that these are two different lists. That's why the check condition returns False.

If you would like to know if two lists have the same elements, turn them into sets and check the condition. Since sets are unordered unlike lists, it will print True confirming that two lists indeed have the same elements.

In [19]:
set1=set(lst1)
set2=set(lst2)

print('This is the first set: {}'.format(set1))
print('This is the second set: {}'.format(set2))

print('----------')

print(set1==set2)


This is the first set: {160, 100, 200, 1800, 140, 1200, 180, 120, 1500}
This is the second set: {160, 100, 1800, 200, 140, 1200, 180, 120, 1500}
----------
True


### List Comprehensions

Here, let's look at an effective way of creating new lists from old ones. Let's say we want to add 5 to each element of the list
lst=[1,2,3,4,5]. Then we could do it by using list comperehsions. 

In the same way, we could create a new tuple by using tuple(); a new set by using set() or a new dictionary by using dict(). For dictionaries, we have to modify the code a little bit.

In [2]:
lst=[1,2,3,4,5]

'''A new list from an old one:'''
new_lst=[x+10 for x in lst]
print(new_lst)

new_lst2=list(x+10 for x in lst)
print(new_lst2)

'''A new tuple from a given list'''
new_tuple=tuple(x+10 for x in lst)
print(new_tuple)

'''A new set from a given list'''
new_set=set(x+10 for x in lst)
print(new_set)

'''A new dictionary from a given list. We have to do this in a different way since dictionaries 
have key, value pairs. So, we have to define keys as well. Notice the difference in the code below:'''
new_dict=dict((x, x+10) for x in lst)

print(new_dict)


[11, 12, 13, 14, 15]
[11, 12, 13, 14, 15]
(11, 12, 13, 14, 15)
{11, 12, 13, 14, 15}
{1: 11, 2: 12, 3: 13, 4: 14, 5: 15}


In [3]:
dict1={'key 1':'apple','key 2':'oranges'}

'''Let's say, we want to create two lists, one for keys of dict1, another for values of dict1. '''

#We first create two empty lists:
keys_list=[]
values_list=[]

#Now, let's iterate over dict1.items() and get the elements we want:
for k, v in dict1.items():
    keys_list.append(k)
    values_list.append(v)

print(keys_list)

print(values_list)


['key 1', 'key 2']
['apple', 'oranges']


## While Statements

A while statement runs if the statement is True and terminates when the statement becomes False. So, we need to clearly define when the loop starts, how it should be modified and when it would stop.

For a complete coverage, you may want to look at the documentation at
https://docs.python.org/3/reference/compound_stmts.html#the-while-statement

In [4]:
# the following code first print(x), then reduce it by 1 and if x<5 it terminates. If we don't have a clear expression of
#how it goes (like reducing by 1) and when it stops, it would run infinitely many times.

lst=[1,2,3,4,5,6,7,8,9,10]

x=9 #you have to give a starting point. If this number is not in lst, the code doesn't start becase of the while statement.

while x in lst: 
    print(x) 
    x=x-1
    if x<5:
        break
        
        
    

9
8
7
6
5
